File: src/csharp/Foundgine.Core/Semantic/SemanticGraphMetadata.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic;

public sealed record SemanticTraversalOrigin(
    EntityId SourceEntity,
    IReadOnlyList<RelationshipId> Path);

public sealed record SemanticIntentOrigin(
    string Operation,
    string? Source = null);

public enum SemanticExpectedCardinality : byte
{
    Unknown,
    One,
    Many
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticGraphMetadata.java
package com.foundgine.core.semantic;
import com.foundgine.core.abstractions.*;
import java.util.List;
public final class SemanticGraphMetadata {
    private SemanticGraphMetadata() {}
    public record TraversalOrigin(EntityId sourceEntity, List<RelationshipId> path) { public TraversalOrigin { path=List.copyOf(path); } }
    public record IntentOrigin(String operation, String source) { public IntentOrigin(String operation) { this(operation,null); } }
    public enum ExpectedCardinality { UNKNOWN, ONE, MANY }
}
